### PR TITLE
renewals: add renew buttons for patrons checked-out items

### PR DESCRIPTION
**Pull request by BadrAly**
* Displays renew button when renewal of checked-out item is possible.
* Hides renew button when circulation policies do not permit the renewal action.
* Increases units testing coverage.
* Corrects indentation of the patron_profile Jinja template.

Co-Authored-by: Aly Badr <aly.badr@rero.ch>

## Why are you opening this PR?

- Renewal by the patron itself
https://tree.taiga.io/project/rero21-reroils/us/350?milestone=247550

## How to test?

- Login as a patron and visit the patron profile page
- For the checked-out tab, if checked-out items can be renewed, the `renew` button will be displayed
- Click the renew button to change the due date of the current check-out transaction

## Code review check list

- [x] Commit message template compliance.
- [x] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [x] Functions docstrings.
- [ ] Unnecessary commited files?

### CODE REVIEW

File: rero_ils/modules/items/api.py [hunk 1]
```diff
@@ -547,10 +547,35 @@ def last_location_pid(self):
             return loan_location_pid
         return self.location_pid
 
+    def can_extend(self, loan):
+        """Checks if an item can be renewed for the borrowed patron."""
+        from ..circ_policies.api import CircPolicy
```
Comment (jma) [inline rero_ils/modules/items/api.py]: Is this can be placed at the top of the file?

File: rero_ils/modules/patrons/views.py [hunk 2]
```diff
@@ -102,17 +104,34 @@ def profile(viewcode):
     patron = Patron.get_patron_by_user(current_user)
     if patron is None:
         raise NotFound()
+    if request.method == 'POST':
+        loan = Loan.get_record_by_pid(request.values.get('loan_pid'))
+        item = Item.get_record_by_pid(loan.get('item_pid'))
+        data = {
+            'item_pid': item.pid, 'pid': request.values.get('loan_pid'),
```
Comment (jma) [inline rero_ils/modules/patrons/views.py]: can you put one key per line? It will be more readable.

File: rero_ils/modules/patrons/views.py [hunk 3]
```diff
@@ -102,17 +104,34 @@ def profile(viewcode):
...
+            'transaction_location_pid': item.location_pid
+        }
+        try:
+            item.extend_loan(**data)
+            flash(_('The item %(item_id)s has been renewed.',
+                    item_id=item.pid), 'success')
+        except Exception:
+            flash(_('Error during the renewal of the item %(item_id)s.',
```
Comment (jma) [inline rero_ils/modules/patrons/views.py]: Does it need a rollback?

File: rero_ils/modules/patrons/views.py [hunk 4]
```diff
@@ -102,17 +104,34 @@ def profile(viewcode):
...
+                    item_id=item.pid), 'danger')
+
     loans = get_loans_by_patron_pid(patron.pid)
     checkouts = []
     requests = []
     if loans:
```
Comment (jma) [inline rero_ils/modules/patrons/views.py]: you can remove this line

File: rero_ils/modules/patrons/views.py [hunk 5]
```diff
@@ -102,17 +104,34 @@ def profile(viewcode):
...
         for loan in loans:
             item_pid = loan.get('item_pid')
-            item = Item.get_record_by_pid(item_pid).replace_refs()
-            document = Document.get_record_by_pid(item['document']['pid'])
+            item = Item.get_record_by_pid(item_pid)
+            document = Document.get_record_by_pid(
+                item.replace_refs()['document']['pid'])
             loan['document_title'] = document['title']
             loan['item_call_number'] = item['call_number']
             if loan['state'] == 'ITEM_ON_LOAN':
```
Comment (jma) [inline rero_ils/modules/patrons/views.py]: is this test required?

File: rero_ils/modules/items/api.py (hunk 1 above)
Comment (BadrAly) [reply rero_ils/modules/items/api.py]: Done

File: rero_ils/modules/patrons/views.py (hunk 2 above)
Comment (BadrAly) [reply rero_ils/modules/patrons/views.py]: done

File: rero_ils/modules/patrons/views.py (hunk 3 above)
Comment (BadrAly) [reply rero_ils/modules/patrons/views.py]: not needed

File: rero_ils/modules/patrons/views.py (hunk 4 above)
Comment (BadrAly) [reply rero_ils/modules/patrons/views.py]: done

File: rero_ils/modules/patrons/views.py (hunk 5 above)
Comment (BadrAly) [reply rero_ils/modules/patrons/views.py]: yes

File: rero_ils/modules/items/api.py [hunk 6]
```diff
@@ -547,10 +548,33 @@ def last_location_pid(self):
             return loan_location_pid
         return self.location_pid
 
+    def can_extend(self, loan):
+        """Checks if an item can be renewed for the borrowed patron."""
```
Comment (iGormilhit) [inline rero_ils/modules/items/api.py]: Is the patron borrowed? :wink: `Checks if the patron as the possibility/has rights/can renew this item.`

File: rero_ils/modules/patrons/views.py [hunk 7]
```diff
@@ -102,28 +104,45 @@ def profile(viewcode):
...
+        loan = Loan.get_record_by_pid(request.values.get('loan_pid'))
+        item = Item.get_record_by_pid(loan.get('item_pid'))
+        data = {
+            'item_pid': item.pid,
+            'pid': request.values.get('loan_pid'),
+            'transaction_location_pid': item.location_pid
+        }
+        try:
+            item.extend_loan(**data)
+            flash(_('The item %(item_id)s has been renewed.',
```
Comment (jma) [inline rero_ils/modules/patrons/views.py]: as you have new translation, please extract the corresponding translations.

Comment (BadrAly) [reply rero_ils/modules/patrons/views.py]: closing this comment, as discussed

**Pull request merged**